builtin/plugins/coin/coin.go

```diff
@@ -213,6 +257,97 @@ func mint(ctx contract.Context, to loom.Address, amount *loom.BigUInt) error {
+	baseAmount.Mul(baseAmount, div)
+	blocksGeneratedPerYear := loom.NewBigUIntFromInt(int64(policy.BlocksGeneratedPerYear))
+	basePercentage := loom.NewBigUIntFromInt(int64(policy.BasePercentage))
+	operator := policy.Operator
+	_, modulus := big.NewInt(ctx.Block().Height).DivMod(big.NewInt(ctx.Block().Height), big.NewInt(int64(policy.BlocksGeneratedPerYear)), big.NewInt(int64(policy.BlocksGeneratedPerYear)))
+	year := blockHeight.Div(blockHeight, blocksGeneratedPerYear)
+	year = year.Add(year, loom.NewBigUIntFromInt(1))
+	var checkMintAmount = &types.BigUInt{
+		Value: *loom.NewBigUIntFromInt(0),
+	}
+	//Checks if minting amount is computed before or it is being computed for first time
+	err = ctx.Get(mintingAmountKey, checkMintAmount)
+	if err != nil && err != contract.ErrNotFound {
+		return err
+	}
+	//Minting Amount per block is set in ctx after minting amount is determined for year
+	if year.Cmp(loom.NewBigUIntFromInt(1)) == 0 {
+		//Determines minting amount at the beginning block of year or at block height at which minting is enabled
+		if modulus.Cmp(big.NewInt(1)) == 0 || err == contract.ErrNotFound {
+			//Minting Amount Computation for starting year
+			amount = baseAmount.Div(baseAmount, blocksGeneratedPerYear)
+			err = ctx.Set(mintingAmountKey, &types.BigUInt{
+				Value: *amount,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		if modulus.Cmp(big.NewInt(1)) == 0 || err == contract.ErrNotFound {
+			//Operator Support to support different inflation ratio patterns for different year
+			if strings.EqualFold("div", operator) {
+				changeRatioDenominator = changeRatioDenominator.Mul(changeRatioDenominator, year)
+				basePercentage = basePercentage.Mul(basePercentage, changeRatioNumerator)
+			} else if strings.EqualFold("exp", operator) {
+				changeRatioNumerator = changeRatioNumerator.Exp(changeRatioNumerator, year, nil)
+				changeRatioDenominator = changeRatioDenominator.Exp(changeRatioDenominator, year, nil)
+				basePercentage = basePercentage.Mul(basePercentage, changeRatioNumerator)
+			} else {
+				return errors.New("Invalid operator - Operator should be div or exp")
+			}
+			var econ Economy
+			err := ctx.Get(economyKey, &econ)
+			if err != nil {
+				return err
+			}
+			//Fetches Total Supply from coin contract and applies inflation ratio on total Supply- Happens at
+			//beginning block for that year or block height at which minting is enabled
+			baseAmount = &econ.TotalSupply.Value
+			inflationforYear := baseAmount.Mul(baseAmount, basePercentage)
+			inflationforYear = inflationforYear.Div(inflationforYear, changeRatioDenominator)
+			inflationforYear = inflationforYear.Div(inflationforYear, loom.NewBigUIntFromInt(100))
+			amount = inflationforYear.Div(inflationforYear, blocksGeneratedPerYear)
+			//Minting Amount per block computed starting from beginning block of that year and saved to avoid re-computations
+			err = ctx.Set(mintingAmountKey, &types.BigUInt{
```

Comment:
Only mint if amount is greater than zero.

Reply:
Also, why are you setting it and then getting it? It's redundant. Just mint and store it.

Reply:
I am already doing this validation here, when minting Amount becomes zero, minting stops and function returns nil  
//Boundary case when minting amount per block becomes zero
	if amount == loom.NewBigUIntFromInt(0) {
		return nil // No more coins to be minted on block creation
	}

Reply:
I am setting base minting amount in context because of following reason - 
BaseMintingAmount remains constant for a year and changes on yearly basis.
So, I am setting it in contract context so that it does not need to be recomputed.
I am storing in contract context so that storage is permanent, in case chain crashes for a short period in a year, base minting amount is already stored. Though it can be recomputed but why will we want to do same computation again :)

Reply:
I am not saying you shouldnt store. I am asking why you are setting it and getting it in same function call.

Reply:
Also, where is the code, that check in code that prevents recomputing of minting amount